### PR TITLE
Add Rally Docker image to release process

### LINKED ISSUE #607: Provide a Rally docker image
In some environments it is troublesome to install all of Rally's dependencies. We should therefore provide a Docker image for Rally during the release process. One aspect that we can also address in a later release are use cases which involve the Rally daemon, like load test drivers that are distributed.

### PR DESCRIPTION
**Pull request by dliappis**
and also add integration tests for the dockerized version of Rally.

~Opening this as draft until we can finalize some details.~

Closes #607

### CODE REVIEW

File: docker/Dockerfiles/Dockerfile-dev [hunk 1]
```diff
@@ -0,0 +1,77 @@
+##############################################################
+# Build stage 0 `builder`:
+# Install Rally from source inside a virtualenv
+################################################################################
```
Comment (danielmitterdorfer) [inline docker/Dockerfiles/Dockerfile-dev]: Very minor nit: The `#` lines have a different length

File: docker/Dockerfiles/Dockerfile-release [hunk 2]
```diff
@@ -0,0 +1,42 @@
+FROM python:3.7.3-slim
+ARG RALLY_VERSION
+ARG RALLY_LICENSE
+
+ENV RALLY_RUNNING_IN_DOCKER True
+
+RUN apt-get -y update && \
+    apt-get install -y curl git gcc && \
+    apt-get -y upgrade && \
```
Comment (danielmitterdorfer) [inline docker/Dockerfiles/Dockerfile-release]: Could we decrease the size of the image a bit further by cleaning up any downloaded files that `apt-get` has cached locally?

File: docker/bin/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+set -Eeo pipefail
+
+if [[ $1 == *"bash"* || $1 == *"sh"* ]]; then
+    : # noop
+elif [[ "$*" == *"--help"* || "$*" == *"list "* ]]; then
+    set -- esrally "$@" # allow --help, --list commands to be passed as is
```
Comment (danielmitterdorfer) [inline docker/bin/entrypoint.sh]: Did you mean to write `list` instead of `--list`?  Also, why does this need special treatment? What about other subcommands like `compare`?

File: integration-test.sh [hunk 4]
```diff
@@ -395,6 +395,54 @@ function test_proxy_connection {
     restore_rally_log
 }
 
+function docker_compose_dev {
+    if [[ "$1" == "up" ]]; then
+        docker-compose -f docker/docker-compose-tests.yml up --abort-on-container-exit
```
Comment (danielmitterdorfer) [inline integration-test.sh]: Should we mention `docker` and `docker-compose` as [prerequisites for development in our docs](https://esrally.readthedocs.io/en/stable/developing.html#prerequisites)?

File: integration-test.sh [hunk 5]
```diff
@@ -395,6 +395,54 @@ function test_proxy_connection {
...
+    elif [[ "$1" == "down" ]]; then
+        docker-compose -f docker/docker-compose-tests.yml down -v
+    fi
```
Comment (danielmitterdorfer) [inline integration-test.sh]: Do you think it makes sense to add an `else` branch that raises an error?

File: integration-test.sh [hunk 6]
```diff
@@ -395,6 +395,54 @@ function test_proxy_connection {
...
+}
+
+function test_docker_image {
+    if ! docker_is_running; then
```
Comment (danielmitterdorfer) [inline integration-test.sh]: I wonder whether we should turn this into a hard requirement now,.

File: integration-test.sh [hunk 7]
```diff
@@ -395,6 +395,54 @@ function test_proxy_connection {
...
+
+    # Build the docker image
+    docker build -t elastic/rally:${RALLY_VERSION} --build-arg RALLY_VERSION --build-arg RALLY_LICENSE -f docker/Dockerfiles/Dockerfile-dev $PWD
+
+    export TEST_COMMAND="--pipeline=benchmark-only --test-mode --track=geonames --challenge=append-no-conflicts-index-only --target-hosts=es01:9200"
+    info "Testing Rally docker image using parameters: ${TEST_COMMAND}"
+    docker_compose_dev up
+    docker_compose_dev down
+
+    # --list should work
```
Comment (danielmitterdorfer) [inline integration-test.sh]: Nit: `list`, not `--list`

File: docker/docker-compose-tests.yml [hunk 8]
```diff
@@ -0,0 +1,38 @@
+version: '2.2'
+services:
+  rally:
+    image: elastic/rally:${RALLY_VERSION}
+    container_name: rally
+    command: ${TEST_COMMAND}
```
Comment (danielmitterdorfer) [inline docker/docker-compose-tests.yml]: I assume this is blocking and we won't run into a race condition in our tests where we basically do a `docker-compose up` followed by a `docker-compose down`?

File: release-docker.sh [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+function docker_compose {
+    if [[ "$1" == "up" ]]; then
+        docker-compose -f docker/docker-compose-tests.yml up --abort-on-container-exit
+    elif [[ "$1" == "down" ]]; then
+        docker-compose -f docker/docker-compose-tests.yml down -v
+    fi
+}
+
+function test_docker_image {
```
Comment (danielmitterdorfer) [inline release-docker.sh]: Could we delegate this testing maybe to `integration-test.sh`, e.g. by calling it with a parameter? I think it would be ok to assume that this workflow is only called by a Makefile target.

File: release-docker.sh [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    docker_compose up
+    docker_compose down
+
+
+    unset TEST_COMMAND
+}
+
+function push_failed {
+    echo "Error while pushing Docker image. Did you \`docker login\`?"
+
```
Comment (danielmitterdorfer) [inline release-docker.sh]: Nit: Trailing new line

File: docker/Dockerfiles/Dockerfile-release (hunk 2 above)
Comment (dliappis) [reply docker/Dockerfiles/Dockerfile-release]: As mentioned in https://docs.docker.com/develop/develop-images/#run (need to scroll further down to APT):

> Official Debian and Ubuntu images automatically run apt-get clean, so explicit invocation is not required.

The image we are basing on (python:3.7.3-slim) is based on the official debian stretch (https://github.com/docker-library/python/blob/4df129649329186fb42c9c33215b41d82bc0e58e/3.7/stretch/slim/Dockerfile#L7) so what we are doing here (i.e. `rm -rf /varl/ib/apt/lists/*`) should be enough.

File: docker/bin/entrypoint.sh (hunk 3 above)
Comment (dliappis) [reply docker/bin/entrypoint.sh]: Good catch left over from previous logic. We don't need special handling for parameters any more, only the "syntactic sugar" of prefixing `esrally` if not specified. Will address.

File: docker/docker-compose-tests.yml (hunk 8 above)
Comment (dliappis) [reply docker/docker-compose-tests.yml]: Correct, docker-compose will return control only when all services are up and satisfy the declared dependencies.

**Comment by dliappis** [comment]
@danielmitterdorfer I addressed the ^^ comments and also added documentation (both regular and release).

File: docs/docker.rst [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    -------------------------------
+
+
+As a next step, we assume that Elasticsearch is running on ``es01:9200`` and is accessible from the host where you are running the Rally Docker image.
+Run the ``nyc_taxis`` track in ``test-mode`` using::
+
+    $ docker run elastic/rally --track=nyc_taxis --test-mode --pipeline=benchmark-only --target-hosts=es01:9200
+
+
+:NOTE: We didn't need to explicitly specify ``esrally`` as we'd normally do in a normal CLI invocation; the entrypoint in the Docker image does this automatically.
```
Comment (danielmitterdorfer) [inline docs/docker.rst]: This is rendered differently than in the rest of our docs. We should use this here:

```rst
.. note::
    We didn't need to explicitly specify ``esrally`` as we'd normally do in a normal CLI invocation; the entrypoint in the Docker image does this automatically.
```

File: docs/docker.rst [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    myrally/benchmarks/races/
+    └── 2019-06-05-14-03-44
+        └── race.json
+
+    1 directory, 1 file
+
+
+In case you forgot to bind mount a directory, the Rally Docker image will create an `anonymous volume <https://success.docker.com/article/different-types-of-volumes>`_ for ``/rally/.rally`` to ensure logs and results get persisted even after the container has terminated.
+
+For example, after executing our earlier quickstart example ``docker run elastic/rally --track=nyc_taxis --test-mode --pipeline=benchmark-only --target-hosts=es01:9200``, ``docker volume ls`` shows a volume:::
```
Comment (danielmitterdorfer) [inline docs/docker.rst]: Nit: This line ends with three colons but it only requires two.

File: docs/docker.rst [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+Specifics about the image
+-------------------------
+
+Rally runs as user ``1000`` and its files are installed with uid:gid ``1000:0`` (to support `Arbitrary User IDs <https://docs.openshift.com/enterprise/3.1/creating_images/guidelines.html>`_).
+
+Extending the Docker image
+--------------------------
+
+Creating your own customized Docker image is simple. You can start with a ``Dockerfile`` as simple as::
```
Comment (danielmitterdorfer) [inline docs/docker.rst]: We should probably rephrase this to avoid calling it "simple"? ([rationale](https://developers.google.com/style/tone)). How about this?

> You can also create your own customized Docker image on top of the existing one. The example below shows how to get started::

File: docs/install.rst [hunk 14]
```diff
@@ -66,6 +66,11 @@ We recommend that you use `Homebrew <https://brew.sh/>`_::
 
     brew install python3
 
+**Docker**
```
Comment (danielmitterdorfer) [inline docs/install.rst]: This is in the wrong section (prerequisites)? I think this should go in the section `Installing Rally` just below `PyEnv Install`?

File: docs/docker.rst [hunk 15]
```diff
@@ -0,0 +1,162 @@
+Running Rally with Docker
+=========================
+
+Rally is also available as a `Docker image <https://hub.docker.com/r/elastic/rally>`_.
```
Comment (danielmitterdorfer) [inline docs/docker.rst]: Remove the `also`?

File: docs/docker.rst [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+
+.. _docker_limitations:
+
+Limitations
+-----------
+
+The following Rally functionality isn't supported or isn't recommended when using the Docker image:
```
Comment (danielmitterdorfer) [inline docs/docker.rst]: I think we can scratch the part `or isn't recommended` as we *actively* prohibit this.

File: docs/install.rst (hunk 14 above)
Comment (dliappis) [reply docs/install.rst]: Right thanks for spotting. Fixed in https://github.com/elastic/rally/pull/702/commits/af1a12402329af7bf27343bc33d63d3fef7abd4b

File: docs/docker.rst (hunk 13 above)
Comment (dliappis) [reply docs/docker.rst]: Addressed in https://github.com/elastic/rally/pull/702/commits/af1a12402329af7bf27343bc33d63d3fef7abd4b

File: docs/docker.rst (hunk 12 above)
Comment (dliappis) [reply docs/docker.rst]: Addressed in https://github.com/elastic/rally/pull/702/commits/af1a12402329af7bf27343bc33d63d3fef7abd4b

Comment (dliappis) [reply docs/docker.rst]: Addressed in https://github.com/elastic/rally/pull/702/commits/af1a12402329af7bf27343bc33d63d3fef7abd4b

File: docs/docker.rst (hunk 11 above)
Comment (dliappis) [reply docs/docker.rst]: Addressed in https://github.com/elastic/rally/pull/702/commits/af1a12402329af7bf27343bc33d63d3fef7abd4b

File: docs/docker.rst [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+    Configuration successfully written to /rally/.rally/rally.ini. Happy benchmarking!
+
+    More info about Rally:
+
+    * Type esrally --help
+    * Read the documentation at https://esrally.readthedocs.io/en/latest/
+    * Ask a question on the forum at https://discuss.elastic.co/c/elasticsearch/rally
+
+    # now run our benchmark
+    docker run --rm -v $PWD/tmpdir:/rally/.rally elastic/rally --track=nyc_taxis --test-mode --pipeline=benchmark-only --target-hosts=es01:9200
```
[the remark this answers is not in the document]
Comment (dliappis) [reply docs/docker.rst]: Addressed in https://github.com/elastic/rally/pull/702/commits/af1a12402329af7bf27343bc33d63d3fef7abd4b

File: docs/docker.rst [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+If your local bind mount doesn't contain a ``rally.ini`` the container will create one for you during the first run.
+
+Example::
+
+    mkdir myrally
+    sudo chgrp 0 myrally
+
+    # First run will generate the rally.ini
+    docker run --rm -v $PWD/tmpdir:/rally/.rally elastic/rally --track=nyc_taxis --test-mode --pipeline=benchmark-only --target-hosts=es01:9200
```
[the remark this answers is not in the document]
Comment (dliappis) [reply docs/docker.rst]: Addressed in https://github.com/elastic/rally/pull/702/commits/af1a12402329af7bf27343bc33d63d3fef7abd4b

File: docs/docker.rst (hunk 16 above)
Comment (dliappis) [reply docs/docker.rst]: Addressed in https://github.com/elastic/rally/pull/702/commits/af1a12402329af7bf27343bc33d63d3fef7abd4b

File: docs/docker.rst (hunk 15 above)
Comment (dliappis) [reply docs/docker.rst]: Addressed in https://github.com/elastic/rally/pull/702/commits/f2e6db89dbd9cde0414581dbfe4a83a808013aa9

**Comment by dliappis** [comment]
@danielmitterdorfer I addressed your comments. If there are no other commends I think we are good to merge?
